### PR TITLE
Fix: Memory.store_n and Helpers.slice_data for border cases

### PR DESCRIPTION
**Pull request by thomas-quadratic**
## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Adding test cases to the opcode 0x39 led to 2 failures. The failures have been identified to be linked to corner cases in Memory.store_n and Helpers.slice_data. Such a corner case is for example when taking a slice with offset out of bound.

Issue Number: related to 98

## What is the new behavior?

Tests for the CODECOPY (0x39) opcode have been extended with more cases taking different code-paths.

Memory.store_n and Helpers.slice_data have been reimplemented to cover all cases.

Cases for slice_data:

  1. slice fully overlaps with data:  data_offset + slice_len <= data_len
  2. slice partially overlaps with data: data_offset < data_len < data_offset + slice_len
  3. slice does not overlap with data: data_len <= data_offset

For Memory.store_n, the new memory is composed of 3 parts:

  1. Head: current bytes before offset. Might be zero padded in case len < offset
  2. New elements: always inserted between offset and offset + element_len
  3. Tail: current bytes after the new elements. Might be zero padded in case total memory length is not a multiple of 32

### CODE REVIEW

File: src/utils/utils.cairo [hunk 1]
```diff
@@ -207,47 +207,23 @@ namespace Helpers {
...
+        let is_non_empty: felt = is_le_felt(data_offset, data_len);
+        let max_len: felt = (data_len - data_offset) * is_non_empty;
         let is_within_bound: felt = is_le_felt(slice_len, max_len);
         let len = max_len + (slice_len - max_len) * is_within_bound;
-        memcpy(dst=new_array, src=array + offset, len=len);
-        fill(arr=new_array + len, value=0, length=slice_len - len);
-        return new_array;
+
+        memcpy(dst=new_data, src=data + data_offset, len=len);
+        fill(arr=new_data + len, value=0, length=slice_len - len);
```
Comment (Eikix) [inline src/utils/utils.cairo:225]: So for an out of bounds offset, with slice_len = x, this function will return an array of length x, filled with zeros?

**Review by Eikix — APPROVED** [review]

File: src/utils/utils.cairo (hunk 1 above)
Comment (thomas-quadratic) [reply src/utils/utils.cairo:225]: Yes! Offset out of bounds means all 0's. I am not sure if there are some opcodes for which the zero-padding rule is not applicable, but on a first look, it seems to be consistent through out the EVM. Arguably, we could separate slicing and padding if ever needed.
